### PR TITLE
Initial Lima backend implementation

### PR DESCRIPTION
**Pull request by mook-as**
This switches to using Lima on macOS.  Note that kim is currently broken: it assumes that we can connect to the node at its IP address, but this is not (yet) the case for Lima.  Normal Kubernetes operations (at least, getting node info) should work correctly.

This is for #380 but would not finish it completely.

### CODE REVIEW

**Review by ericpromislow — COMMENTED** [review]
This doesn't work out of the box. `lima.log` ends with these messages:
```
time="2021-07-07T11:22:14-07:00" level=info msg="Using the existing instance \"rancher-desktop\""
time="2021-07-07T11:22:14-07:00" level=info msg="[hostagent] Starting QEMU (hint: to watch the boot progress, see \"/Users/ericp/Library/State/rancher-desktop/lima/rancher-desktop/serial.log\")"
time="2021-07-07T11:22:14-07:00" level=info msg="SSH Local Port: 60020"
time="2021-07-07T11:22:14-07:00" level=info msg="[hostagent] Waiting for the essential requirement 1 of 4: \"ssh\""
…
time="2021-07-07T11:32:08-07:00" level=info msg="[hostagent] Waiting for the essential requirement 1 of 4: \"ssh\""
time="2021-07-07T11:32:08-07:00" level=info msg="[hostagent] Waiting for the essential requirement 2 of 4: \"sshfs binary to be installed\""
time="2021-07-07T11:32:14-07:00" level=fatal msg="did not receive an event with the \"running\" status"
```

at which point `background.log` reports:
```
==> background.log <==
Kubernetes was unable to start: Error: /Users/ericp/workspace/rancher/desktop/resources/darwin/lima/bin/limactl exited with code 1
    at ChildProcess.<anonymous> (/Users/ericp/workspace/rancher/desktop/dist/app/background.js:25719:20)
    at ChildProcess.emit (events.js:315:20)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:277:12)
```

Maybe related: trying to run `.../limactl start` fails with this output:
```
INFO[0013] Downloading "https://github.com/containerd/nerdctl/releases/download/v0.10.0/nerdctl-full-0.10.0-linux-amd64.tar.gz" (sha256:5ba3d476707d510fe3ca3928e9cda5d0b4ce527d42b343404c92d563f82ba967)
172.58 MiB / 172.58 MiB [-----------------------------------] 100.00% 8.56 MiB/s
INFO[0034] Downloaded "nerdctl-full-0.10.0-linux-amd64.tar.gz"
INFO[0035] Attempting to download the image from "~/Downloads/hirsute-server-cloudimg-amd64.img"
INFO[0035] Attempting to download the image from "https://cloud-images.ubuntu.com/hirsute/current/hirsute-server-cloudimg-amd64.img"
551.62 MiB / 551.62 MiB [----------------------------------] 100.00% 20.54 MiB/s
INFO[0063] Downloaded image from "https://cloud-images.ubuntu.com/hirsute/current/hirsute-server-cloudimg-amd64.img"
FATA[0063] failed to run [qemu-img create -f qcow2 -b /Users/ericp/.lima/default/basedisk /Users/ericp/.lima/default/diffdisk 107374182400]: "": exec: "qemu-img": executable file not found in $PATH
```

**Comment by jandubois** [comment]
> Maybe related: trying to run `.../limactl start` fails with this output:

@ericpromislow You need to put the `bin` directory in your `PATH`:

```
PATH=$PWD/resources/darwin/lima/bin resources/darwin/lima/bin/limactl start
```

Your problem might be related to either your `~/.ssh/*.pub` files or `~/.ssh/config` settings. Just for a test, could you rename `~/.ssh` to something else and try again?

**Comment by mattfarina** [comment]
I was able to get everything working except `kim`. `kim` is looking up the IP to use and gets a 192.168.x.x IP to try and communicate with. That's not accessible. We'll have to look into this further.

**Comment by ericpromislow** [comment]
Looks like it's an environment issue:
```
FATA[0004] failed to run [qemu-img create -f qcow2 -b /Users/ericp/.lima/default/basedisk /Users/ericp/.lima/default/diffdisk 107374182400]: "dyld: Library not loaded: @executable_path/../opt/libssh/lib/libssh.4.dylib\n  Referenced from: /usr/local/bin/qemu-img\n  Reason: image not found\n": signal: abort trap
```

Adding this for macos is not that straightforward. libssl.org and libssl2.org are two separate projects, and neither appears to provide this version. Brew doesn't know either. I could download an app that uses it and put that in the path, but that's an awful idea

**Review by mattfarina — COMMENTED** [review]
With the exception of kim issues, the PR looks good. We need to work out the kim issues though.

**Comment by jandubois** [comment]
> Adding this for macos is not that straightforward. libssl.org and libssl2.org are two separate projects, and neither appears to provide this version. Brew doesn't know either. I could download an app that uses it and put that in the path, but that's an awful idea

@ericpromislow This should not be necessary; all the libs are bundled with RD:

```console
$ ls -l resources/darwin/lima/opt/libssh/lib
total 800
-rw-r--r--  1 jan  staff  403128  6 Jul 09:52 libssh.4.8.6.dylib
lrwxr-xr-x  1 jan  staff      18  6 Jul 09:52 libssh.4.dylib -> libssh.4.8.6.dylib
```

Did you re-arrange anything under `resources/darwin/lima`?

**Comment by ericpromislow** [comment]
> > Adding this for macos is not that straightforward. libssl.org and libssl2.org are two separate projects, and neither appears to provide this version. Brew doesn't know either. I could download an app that uses it and put that in the path, but that's an awful idea
> 
> @ericpromislow This should not be necessary; all the libs are bundled with RD:
> 
> ```
> $ ls -l resources/darwin/lima/opt/libssh/lib
> total 800
> -rw-r--r--  1 jan  staff  403128  6 Jul 09:52 libssh.4.8.6.dylib
> lrwxr-xr-x  1 jan  staff      18  6 Jul 09:52 libssh.4.dylib -> libssh.4.8.6.dylib
> ```
> 
> Did you re-arrange anything under `resources/darwin/lima`?

This all happened by trying to run lima in isolation. As noted elsewhere, it finally started up while running in RD.

File: src/k8s-engine/lima.ts [hunk 1]
```diff
@@ -0,0 +1,572 @@
...
+  }
+
+  get memory(): Promise<number> {
+    return (async() => {
+      return Math.round(((await this.currentConfig)?.memory || 0) / 1024 / 1024 / 1024);
+    })();
+  }
+
+  /** Get the IPv4 address of the VM, assuming it's already up */
+  get ipAddress(): Promise<string | undefined> {
```
Comment (AkihiroSuda) [inline src/k8s-engine/lima.ts:230]: Why not use `ip` command

**Comment by AkihiroSuda** [comment]
> I was able to get everything working except kim. kim is looking up the IP to use and gets a 192.168.x.x IP to try and communicate with. That's not accessible. We'll have to look into this further.

Do you expect the issue to be fixed on kim side or on Lima/QEMU side?

Ideally we should use [the VMnet patch for QEMU](https://lore.kernel.org/qemu-devel/20210617143246.55336-1-yaroshchuk2000@gmail.com/), but I'm not sure when the patch will get merged in the upstream.
I guess we could use TAP or some VPN instead, though, but they will require sudo 😞

**Comment by jandubois** [comment]
> Do you expect the issue to be fixed on kim side or on Lima/QEMU side?

We intend to implement a workaround in kim until lima supports the vmnet framework:

* kim will be updated to support a hidden option to `kim builder install` to add an annotation on the builder service.
* When kim connects to a builder service with this annotation, it will not use the service IP address, but connect to the same port on localhost.

This should generally work as long as `kim` is invoked directly from the host. It will not work if it is invoked from inside another VM or a container.

That seems like a reasonable limitation. As an extension to this idea we may store the host IP address in the kim annotation and have RD request lima to bind the builder port to `0.0.0.0`. This should provide a routable path to the builder service, but it will expose the service to the outside of the dev machine. So this will have to be configurable. We have not decided yet, if we want to implement this extension, or if we just want to wait until vmnet support becomes available.

Further thoughts:

1. RD ships with qemu binaries, so requiring a patched version is not really a problem. Not sure how much effort implementing vmnet support is in lima itself.
2. We ideally want the clusterip information for k8s services to be correct, so users can build applications that rely on this. So even if we can work around the problems for kim, it does not really solve all the problems.
3. We can probably use the "bind to 0.0.0.0" workaround for the traefik ingress, but again showing the right IP address for the service endpoint would be preferable.

**Comment by AkihiroSuda** [comment]
Looks like signing QEMU binary with `com.apple.vm.networking` requires agreement with Apple https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_vm_networking 

So Lima isn't highly likely to be able to support VMnet... 😞 
https://github.com/AkihiroSuda/lima/issues/81

At least, it will require running a sudo helper 😞

**Comment by jandubois** [comment]
> Experimental `vmnet.framework` support for unpatched rootless QEMU: https://github.com/AkihiroSuda/vde_vmnet

Thank you! That looks promising!

> This `vde_vmnet` does not require QEMU to run as the root user, though `vde_vmnet` itself has to run as the root

Yes, that is too bad; I was hoping we could avoid needing `sudo` completely. However, this is no worse situation than with `hyperkit`, which also need to run as `root` for networking reasons.

On the bright side, having a `sudo` helper opens the door to (optionally) providing NFS mounts for performance reasons. I would like to not do this, but we'll have to see how the current sshfs setup holds up to various use cases.

**Comment by jandubois** [comment]
> Looks like signing QEMU binary with `com.apple.vm.networking` requires agreement with Apple https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_vm_networking

And the entitlement [was created for App Store apps](https://developer.apple.com/forums/thread/116870):

>However, I have to set some expectations here: `com.apple.vm.networking` was intended to allow VM apps to ship in the Mac App Store (where privilege escalation is outlawed) and thus it’s going to be a hard sell for non-app products shipping outside the store.
-- Quinn “The Eskimo!” (Apple Developer Relations, Developer Technical Support, Core OS/Hardware)

So it sounds like even SUSE/Rancher would have a hard time getting it; we would be told to just run as `root` instead.

**Comment by mook-as** [comment]
Updated lima to make it use localhost — that works now.

**Comment by ericpromislow** [comment]
Images still aren't showing up.  I get these messages in `k3s.log` (on macos):
```
E0726 18:18:47.741403    2858 remote_runtime.go:228] "CreateContainer in sandbox from runtime service failed" err="rpc error: code = Unknown desc = failed to generate container \"416864a05299d13b1faa81fe652974fb7c805a09a508ef39e8e8096156d80b8a\" spec: failed to generate spec: path \"/tmp\" is mounted on \"/\" but it is not a shared mount"
...
E0726 18:18:47.744262    2858 remote_runtime.go:228] "CreateContainer in sandbox from runtime service failed" err="rpc error: code = Unknown desc = failed to generate container \"0591ec86a275b1d54dd8830217387b384b70d490cd43e961d926096e8e67fda1\" spec: failed to generate spec: path \"/var/lib/buildkit\" is mounted on \"/\" but it is not a shared mount"
…
0726 18:18:47.744565    2858 kuberuntime_manager.go:864] container &Container{Name:agent,Image:rancher/kim:v0.1.0-beta.4,Command:[kim --debug agent],Args:[--agent-port=1233 --buildkit-socket=unix:///run/buildkit/buildkitd.sock --buildkit-port=1234 --containerd-socket=/run/k3s/containerd/containerd.sock --tlscacert=/certs/ca/tls.crt --tlscert=/certs/server/tls.crt --tlskey=/certs/server/tls.key],WorkingDir:,Ports:[]ContainerPort{ ... }

CreateContainerError: failed to generate container "0591ec86a275b1d54dd8830217387b384b70d490cd43e961d926096e8e67fda1" spec: failed to generate spec: path "/var/lib/buildkit" is mounted on "/" but it is not a shared mount

E0726 18:18:47.744690    2858 pod_workers.go:190] "Error syncing pod, skipping" err="[failed to \"StartContainer\" for \"buildkit\" with CreateContainerError: \"failed to generate container \\\"416864a05299d13b1faa81fe652974fb7c805a09a508ef39e8e8096156d80b8a\\\" spec: failed to generate spec: path \\\"/tmp\\\" is mounted on \\\"/\\\" but it is not a shared mount\", failed to \"StartContainer\" for \"agent\" with CreateContainerError: \"failed to generate container \\\"0591ec86a275b1d54dd8830217387b384b70d490cd43e961d926096e8e67fda1\\\" spec: failed to generate spec: path \\\"/var/lib/buildkit\\\" is mounted on \\\"/\\\" but it is not a shared mount\"]"
```

Those character escapes on the last message look like the quotation marks are being picked up as part of the pathname.

**Comment by mook-as** [comment]
I don't understand why you're seeing those issues but I'm not… but hopefully a315a1368de9c5d228a591a21ebb4d24a6bae9b3 will do something about; it's copied from code you did for WSL.

If that still gives you problems, please be sure to delete the lima VM (`limactl delete rancher-desktop` with the correct `LIMA_HOME` setup); that shouldn't affect actual users because they wouldn't have had an older lima VM lying around.

**Comment by mattfarina** [comment]
When I tried to run `kim build .` in the root of [kubewarden-controller](https://github.com/kubewarden/kubewarden-controller) I got the error:

> failed to compute cache key: failed to copy: failed to send write: write /var/lib/rancher/k3s/agent/containerd/io.containerd.content.v1.content/ingest/afe6eb9430538e21dcc29ddeadbd89d85a4bfc29225efc321778383d13fb391c/data: no space left on device: unknown

With a smaller scratch image I had no issues. Under hyperkit I'd had no issue with building this image. It is one of my demo examples.

**Comment by mattfarina** [comment]
More detail... going into the VM and running `df` reveals...

```
Filesystem           1K-blocks      Used Available Use% Mounted on
devtmpfs                 10240         0     10240   0% /dev
shm                    1502676         0   1502676   0% /dev/shm
/dev/sr0                 62464     62464         0 100% /media/cdrom
tmpfs                  1502676   1357604    145072  90% /
tmpfs                   300536      1000    299536   0% /run
/dev/loop0               12928     12928         0 100% /.modloop
/dev/disk/by-label/cidata
                          6822      6822         0 100% /mnt/lima-cidata
/dev/disk/by-label/data-volume
                     102686648     68896  97358540   0% /mnt/data
cgroup_root              10240         0     10240   0% /sys/fs/cgroup
:/Users/mfarina/Library/Caches/rancher-desktop/k3s
                     1048576000         0 1048576000   0% /Users/mfarina/Library/Caches/rancher-desktop/k3s
df: /var/lib/kubelet/pods/db197a29-7be7-48cc-9d39-a86373f1a316/volumes/kubernetes.io~secret/certs-ca: Permission denied
df: /var/lib/kubelet/pods/db197a29-7be7-48cc-9d39-a86373f1a316/volumes/kubernetes.io~secret/certs-server: Permission denied
df: /var/lib/kubelet/pods/db197a29-7be7-48cc-9d39-a86373f1a316/volumes/kubernetes.io~projected/kube-api-access-2nqzz: Permission denied
overlay                1502676   1357604    145072  90% /run/k3s/containerd/io.containerd.runtime.v2.task/k8s.io/7ba507d24ed0708e1bf2ab127cc7e0cd994768af54827c820cceeb911f9fb7af/rootfs
df: /var/lib/kubelet/pods/7b68fa87-0de4-4932-bfab-7e7b39385bd4/volumes/kubernetes.io~projected/kube-api-access-wds7x: Permission denied
shm                      65536         0     65536   0% /run/k3s/containerd/io.containerd.grpc.v1.cri/sandboxes/5dcfaaeaf86341b25964dc7e51a111211adda1f647e680770f228fec1ad99805/shm
overlay                1502676   1357604    145072  90% /run/k3s/containerd/io.containerd.runtime.v2.task/k8s.io/5dcfaaeaf86341b25964dc7e51a111211adda1f647e680770f228fec1ad99805/rootfs
overlay                1502676   1357604    145072  90% /run/k3s/containerd/io.containerd.runtime.v2.task/k8s.io/cf0b145550350eb21c697b65541da707995be749d8e52704d7ae9f7ecb720d14/rootfs
df: /var/lib/kubelet/pods/a6ed86b6-97fa-492e-9650-52f947bfd70f/volumes/kubernetes.io~projected/kube-api-access-kclnf: Permission denied
shm                      65536         0     65536   0% /run/k3s/containerd/io.containerd.grpc.v1.cri/sandboxes/cbb47a7d027c064613a25fb34f4573ecdcd690a5f37913c68f85db5b05215065/shm
overlay                1502676   1357604    145072  90% /run/k3s/containerd/io.containerd.runtime.v2.task/k8s.io/cbb47a7d027c064613a25fb34f4573ecdcd690a5f37913c68f85db5b05215065/rootfs
df: /var/lib/kubelet/pods/44278493-1292-4d78-9f83-03d82f29de66/volumes/kubernetes.io~projected/kube-api-access-d5mmc: Permission denied
shm                      65536         0     65536   0% /run/k3s/containerd/io.containerd.grpc.v1.cri/sandboxes/ef7fa674e617eb3eba047c54c1ef7d8452f9fa3591a5cfc0c521181672080f14/shm
overlay                1502676   1357604    145072  90% /run/k3s/containerd/io.containerd.runtime.v2.task/k8s.io/ef7fa674e617eb3eba047c54c1ef7d8452f9fa3591a5cfc0c521181672080f14/rootfs
overlay                1502676   1357604    145072  90% /run/k3s/containerd/io.containerd.runtime.v2.task/k8s.io/0587669f7cd11096ded470bc2f6c8a6211de573c1283ed6bff7339662cd2892a/rootfs
overlay                1502676   1357604    145072  90% /run/k3s/containerd/io.containerd.runtime.v2.task/k8s.io/87602b1b555c418efba923e036dcbc88f983b1c9f47559f2416fadc5267ff033/rootfs
overlay                1502676   1357604    145072  90% /run/k3s/containerd/io.containerd.runtime.v2.task/k8s.io/d96bf5992d3bb60bbb97e3c360ef8c751116c2585b3b9a02dc45ff548fc96cbf/rootfs
df: /var/lib/kubelet/pods/263a6d3f-0790-4f76-abf3-253792707845/volumes/kubernetes.io~projected/kube-api-access-d5v9p: Permission denied
df: /var/lib/kubelet/pods/607ba1c6-d9ec-4186-82a7-679aae76029a/volumes/kubernetes.io~projected/kube-api-access-ttgm2: Permission denied
shm                      65536         0     65536   0% /run/k3s/containerd/io.containerd.grpc.v1.cri/sandboxes/75a7065364ac39fafc64258f1c1c94b8e549e6290cbbba87b09dd4b89f01d487/shm
overlay                1502676   1357604    145072  90% /run/k3s/containerd/io.containerd.runtime.v2.task/k8s.io/75a7065364ac39fafc64258f1c1c94b8e549e6290cbbba87b09dd4b89f01d487/rootfs
shm                      65536         0     65536   0% /run/k3s/containerd/io.containerd.grpc.v1.cri/sandboxes/6ec3aba22d473c2e3ea9eebac56bf3c47abe63aa652b97e878d48f3df6dda925/shm
overlay                1502676   1357604    145072  90% /run/k3s/containerd/io.containerd.runtime.v2.task/k8s.io/b99478ce54fc4dd7bfabf63599512b54c37d2af9294b296601186415a6b6ffdb/rootfs
overlay                1502676   1357604    145072  90% /run/k3s/containerd/io.containerd.runtime.v2.task/k8s.io/6ec3aba22d473c2e3ea9eebac56bf3c47abe63aa652b97e878d48f3df6dda925/rootfs
overlay                1502676   1357604    145072  90% /run/k3s/containerd/io.containerd.runtime.v2.task/k8s.io/3f7231c5f1a2721b4dd4a7999f16481ba42971bd43faabd7657b9efe8c3eaea7/rootfs
overlay                1502676   1357604    145072  90% /run/k3s/containerd/io.containerd.runtime.v2.task/k8s.io/64d9e7092a774fc4bcae0d5e4f97b26cd90967924d13b08b5cd9ff4247a9058f/rootfs
overlay                1502676   1357604    145072  90% /run/k3s/containerd/io.containerd.runtime.v2.task/k8s.io/4b030d26e5bbe25be8386d6b3a97f6b3fa51555d9795a1c4a185f265625ec6f8/rootfs
```

`/var/` does not appear to have enough space as it's part of `/`. This is likely what's causing the issue.

**Comment by mook-as** [comment]
- Dropped the `mount --make-shared`, it's already in the provisioning script.
- Mounted `/var/lib/buildkit` and `/tmp` into the data disk as well; hopefully that will resolve issues with building images.  Opted to _not_ mount all of `/var/lib`, as it already has things like `/var/lib/sudo` at that point, and messing with existing things seem like it might break things.  If we need to, I suppose I can bind mount all of it into `/mnt/data/var/lib/…` first, and the bind mount `/var/lib` back…

File: src/k8s-engine/k8s.ts [hunk 2]
```diff
@@ -165,7 +168,7 @@ export interface KubernetesBackend extends events.EventEmitter {
 export function factory(): KubernetesBackend {
   switch (os.platform()) {
   case 'darwin':
-    return new Hyperkit();
+    return new LimaBackend();
```
Comment (ericpromislow) [inline src/k8s-engine/k8s.ts:171]: Nothing is calling `Hyperkit()` anymore. Can it be removed, or is there a chance we'll need it in the future. And if we don't need it anymore, can we also drop scripts/download/hyperkit.mjs and src/k8s-engine/hyperkit.ts ?

**Review by ericpromislow — APPROVED** [review]

**Comment by mook-as** [comment]
Rebased and fixed to work with the current `main`.  Also dropped hyperkit as requested.  If we need it in the future, `git revert` is easy.

**Comment by mattfarina** [comment]
QA testing I discovered that:

- Using the Factory Reset Button isn't working. I noticed that it does not download k3s again.
- After manually deleting the state, cache, and prefs directories I see a message on the "Kubernetes Settings" screen that says "The cluster must be reset for port change from 0 to 6443." and the Kubernetes cluster does not start I get a pop-up that says, "There was an unknown error starting Kubernetes: Error: /Users/mfarina/code/rancher-sandbox/rd/resources/darwin/lima/bin/limactl exited with code 1"

This started happening when I downgraded from 1.21.3 to 1.18.7. I was testing a downgrade and everything started to break. So I reset Kubernetes and then tried a Factory Reset. Same errors kept happening.

**Comment by mook-as** [comment]
Could you attach `lima.log` (or other logs as appropriate), so we can try to track that down?

**Comment by mattfarina** [comment]
Note, my `kim build` tests work, now.

**Comment by mook-as** [comment]
The port 0 should be fixed in 29ea0a0653429672ac69c05a5273dd0e3a0d3a51 (just suppress all restart notifications when we're starting).

I could not reproduce the error on starting.  I assume factory reset will be done in a separate PR.

**Comment by mattfarina** [comment]
```
❯ cat ~/Library/State/rancher-desktop/logs/lima.log
time="2021-07-27T16:44:44-04:00" level=info msg="Using the existing instance \"rancher-desktop\""
time="2021-07-27T16:44:45-04:00" level=info msg="[hostagent] Starting QEMU (hint: to watch the boot progress, see \"/Users/mfarina/Library/State/rancher-desktop/lima/rancher-desktop/serial.log\")"
time="2021-07-27T16:44:45-04:00" level=info msg="SSH Local Port: 60020"
time="2021-07-27T16:44:45-04:00" level=info msg="[hostagent] QEMU has exited"
time="2021-07-27T16:44:45-04:00" level=fatal msg="exiting, status={Running:false Degraded:false Exiting:true Errors:[] SSHLocalPort:0} (hint: see \"/Users/mfarina/Library/State/rancher-desktop/lima/rancher-desktop/ha.stderr.log\")"
Error starting lima: Error: /Users/mfarina/code/rancher-sandbox/rd/resources/darwin/lima/bin/limactl exited with code 1
    at ChildProcess.<anonymous> (/Users/mfarina/code/rancher-sandbox/rd/dist/app/background.js:30848:27)
    at ChildProcess.emit (events.js:315:20)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:277:12)
```

That's the output of my lima.log file.

**Comment by mook-as** [comment]
I can reproduce the errors starting up if I intentionally make that port busy via `nc -l 127.0.0.1 60020` — that might need to be a follow-up PR?

**Review by mattfarina — APPROVED** [review]
I've been using this for the past couple days while testing hypper. Appears to be working for me.

**Pull request merged**